### PR TITLE
[aws-datastore] three-way merge logic for unconditional mutations

### PR DESCRIPTION
**Pull request by jamesonwilliams**
Add enqueuing rules to ensure that there is always at most one mutation
for a given model ID, in the outbox, at a time.

When there is an already-pending mutation for a given model ID, and
another mutation is requested to be enqueued, these rules are applied:

existing creation incoming creation -> error
existing creation incoming update   -> incoming data, existing ID
existing creation incoming deletion -> delete exiting

existing update   incoming creation -> error
existing update   incoming update   -> incoming data, existing ID
existing update   incoming deletion -> existing ID, type now deletion

existing delete   incoming creation -> error
existing delete   incoming update   -> error
existing delete   incoming deletion -> exiting retained

When there is no existing mutation, a pending mutation is always
enqueued.

Future work will address the scenario where the mutations have
conditional predicates attached to them.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Merger.java [hunk 1]
```diff
@@ -145,4 +155,22 @@
...
         }, failure -> onNotPresent.call());
     }
+
+    /**
+     * The strategy to use while merging. Whether to consider the contents of the mutation
+     * outbox before saving data locally, or, to ignore it.
+     */
+    enum MergeStrategy {
+        /**
+         * When merging, the contents of the mutation outbox will *not* be considered.
```
Comment (jpignata) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Merger.java:165]: I'm not sure I see this branch in the requirements of [JS implementation](https://github.com/aws-amplify/amplify-js/blob/master/packages/datastore/src/sync/merger.ts) -- why do we ignore mutations and clobber on a network update when it seems like the intent is wait until the last mutation is applied?

Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Merger.java:165]: There are four sources of state in this system:
1. User's data, received from DataStoreAPI;
2. Mutation responses, received when user's data is published;
3. Subscription responses, received when any data change occurs on the remote system;
4. Sync responses, received at system startup.

3 and 4 don't have pending mutations, we just get inbound data. 1 generates a pending mutation, but doesn't process it.

2 is unique. When the mutation processor starts processing a pending mutation, it is _guaranteed_ that it is in the outbox (that's where it came from.) So by the time it gets around to merging it, we _expect_ that mutation to be in the outbox. (We haven't removed it, yet.) In fact, "clobber existing" could be more strongly implemented as "demand that it DOES exist, and fail if it doesn't".

File: aws-datastore/src/androidTest/java/com/amplifyframework/datastore/AWSDataStorePluginInstrumentedTest.java [hunk 2]
```diff
@@ -57,19 +61,11 @@
  * which were defined by the schema in:
  * testmodels/src/main/java/com/amplifyframework/testmodels/commentsblog/schema.graphql.
  */
-@Ignore("AWSDataStorePlugin must not refer to Amplify.API - need to update source")
+@Ignore("This test is not reliably passing right now.")
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-datastore/src/androidTest/java/com/amplifyframework/datastore/AWSDataStorePluginInstrumentedTest.java:64]: Same. This is a top concern and priority for me.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/model/SystemModelsProviderFactory.java [hunk 3]
```diff
@@ -43,6 +43,10 @@ public static ModelProvider create() {
         return SimpleModelProvider.instance(
             SYSTEM_MODELS_VERSION,
 
+            // Used to create a persistent queue of mutations that need to be dispatched over
+            // the network.
+            PendingMutation.PersistentRecord.class,
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/model/SystemModelsProviderFactory.java:48]: It did fix a SQL foreign key violation for me. But, I haven't thought deeply about it yet to explain _why_ it fixes a foreign key violation. Some other part of the system expects this thing to be around, early on, though.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java [hunk 4]
```diff
@@ -71,24 +73,41 @@
...
+                return Completable.complete();
+            }
+            boolean itemFailedToProcess = !processOutboxItem(next)
+                .blockingAwait(ITEM_PROCESSING_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (itemFailedToProcess) {
+                return Completable.error(new DataStoreException(
+                    "Failed to process " + next, "Check your internet connection."
+                ));
+            }
+        } while (true);
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java:108]: The `while (true)` looks sketchy. I think its ~okay, though. The interior of the `do { }` says "If the queue is empty, return now. It not empty, process the next. If that fails, return now." So the loop only proceeds when it was not empty, and the last item was successfully processed.

Agree we do need some provisions around what happens when a publication fails, though.

**Pull request merged**